src/machine/instruction.cpp

```diff
@@ -1514,3 +1514,133 @@ TokenizedInstruction::TokenizedInstruction(
+        for (int j = 0; j < ((1 << im->subfield.count) - 1); j++) {
+            gen_instruction_code_and_expected_str(outfile, &im->subclass[j], code, im->subfield.count, im->subfield.offset, j, depth + 1);
+        }
+        return;
+    }
+
+    QString res = im->name;
+    
+    std::unordered_map<std::string, uint32_t> direct_map = {
+        { "ecall", 0b00000000000000000000000001110011 },
+        {  "mret", 0b00110000001000000000000001110011 },
+        {  "sret", 0b00010000001000000000000001110011 }
+    };
+    if (direct_map.count(im->name)) {
+        code = direct_map[im->name];
+    } else {
+
+
+    int val = 1;
+    switch (im->type) {
+        case Instruction::Type::R:
+        case Instruction::Type::AMO:  {
+            code |= 0b00000000000100001000000010000000;
+            break;
+        }
+        case Instruction::Type::I: 
+            code |= 0b00000000000100001000000010000000;
+            break;
+        case Instruction::Type::ZICSR: {
+            code |= 0b00000000000100001000000010000000;
+            break;
+        }
+        case Instruction::Type::S: {
+            code |= 0b00000000000100001000000010000000;
+            break;
+        }
+        case Instruction::Type::B: {
+            code |= 0b00000000000100001000000100000000;
+            val = 2;
+            break;
+        }
+        case Instruction::Type::U: {
+            code |= 0b00000000000000000001000010000000;
+            break;
+        }
+        case Instruction::Type::J: {
+            code |= 0b00000000001000000000000010000000;
+            val = 2;
+            break;
+        }
+        case Instruction::Type::UNKNOWN: {
+            return;
+        }
+    }
+
+    QString next_delim = " ";
+    for (const QString &arg_string : im->args) {
+        res += next_delim;
+        next_delim = ", ";
+        for (int pos = 0; pos < arg_string.size(); pos += 1) {
+            char arg_letter = arg_string[pos].toLatin1();
+            const ArgumentDesc *arg_desc = arg_desc_by_code[(unsigned char)arg_letter];
+            if (arg_desc == nullptr) {
+                res += arg_letter;
+                continue;
+            }
+            switch (arg_desc->kind) {
+            case 'g': {
+                res += "x1";
+                break;
+            }
+            case 'p':
+            case 'a':
+                res += QString::asprintf("0x%d", val);
+                break;
+            case 'o':
+            case 'n': {
+                if (arg_desc->min < 0) {
+                    res += "1";
+                } else {
+                    res += "0x1";
+                }
+                break;
+            }
+            case 'E': {
+                res += "0x1";
+                break;
+            }
+            }
+        }
+    }
+    }
+
+
+    QString test_code = QString::asprintf("QCOMPARE(Instruction(0x%x).to_str(), \"%s\");", code, qPrintable(res));
+    outfile << test_code;
+    if (Instruction(code).to_str() != res) {
```

Comment:
I am not sure why this is here. I would expect the test generation and execution to be separate.

Reply:
Should it be in a separate file as an executable? (for generating test cases)